src/api/syscall.c

```diff
@@ -329,8 +329,13 @@ static exception_t handleInvocation(bool_t isCall, bool_t isBlocking)
         return EXCEPTION_NONE;
     }
 
+    /* The EXCEPTION_NONE_THRESHOLD_RESTART check is necessary,
+     * because for IPC threshold behaviour, we actually do want to set ThreadState_Restart
+     */
     if (unlikely(
-            thread_state_get_tsType(thread->tcbState) == ThreadState_Restart)) {
+            thread_state_get_tsType(thread->tcbState) == ThreadState_Restart
+                && status != EXCEPTION_NONE_THRESHOLD_RESTART
+            )) {
```

Comment:
The original code was already obtuse, but with this change it gets even harder to understand what's happening.

It would make more sense if the code that sets the task to `ThreadState_Restart` would set it to `ThreadState_Running` instead if that's what's wanted (and also call `replyFromKernel_success_empty()`). Then this code can be removed altogether.